docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,612 @@
+	// we don't open a dstFile until we receive our first ABCI message
+	if err := fileutil.IsDirWriteable(writeDir); err != nil {
+		return nil, err
+	}
+	return &FileStreamingService{
+		listeners: listeners,
+		srcChan: listenChan,
+		filePrefix: filePrefix,
+		writeDir: writeDir,
+		marshaller: m,
+		stateCache: make([][]byte, 0),
+	}, nil
+}
+
+// Listeners returns the StreamingService's underlying WriteListeners, use for registering them with the BaseApp
+func (fss *FileStreamingService) Listeners() map[sdk.StoreKey][]storeTypes.WriteListener {
+	return fss.listeners
+}
+
+func (fss *FileStreamingService) ListenBeginBlock(ctx sdk.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) {
+	// NOTE: this could either be done synchronously or asynchronously
+	// create a new file with the req info according to naming schema
+	// write req to file
+	// write all state changes cached for this stage to file
+	// reset cache
+	// write res to file
+	// close file
+}
+
+func (fss *FileStreamingService) ListenEndBlock(ctx sdk.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) {
+	// NOTE: this could either be done synchronously or asynchronously
+	// create a new file with the req info according to naming schema
+	// write req to file
+	// write all state changes cached for this stage to file
+	// reset cache
+	// write res to file
+	// close file
+}
+
+func (fss *FileStreamingService) ListenDeliverTx(ctx sdk.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) {
+	// NOTE: this could either be done synchronously or asynchronously
+	// create a new file with the req info according to naming schema
+	// NOTE: if the tx failed, handle accordingly
+	// write req to file
+	// write all state changes cached for this stage to file
+	// reset cache
+	// write res to file
+	// close file
+}
+
+// Stream spins up a goroutine select loop which awaits length-prefixed binary encoded KV pairs and caches them in the order they were received
+func (fss *FileStreamingService) Stream(wg *sync.WaitGroup, quitChan <-chan struct{}) {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-quitChan:
+				return
+                        case by := <-fss.srcChan:
+                        	append(fss.stateCache, by)
+			}
+		}
+	}()
+}
+```
+
+Writing to a file is the simplest approach for streaming the data out to consumers.
+This approach also provide the advantages of being persistent and durable, and the files can be read directly,
+or an auxiliary streaming services can read from the files and serve the data over a remote interface.
+
+#### Auxiliary streaming service
+
+We will create a separate standalone process that reads and internally queues the state as it is written out to these files
+and serves the data over a gRPC API. This API will allow filtering of requested data, e.g. by block number, block/tx hash, ABCI message type,
```

Comment:
Is this for clients or caching to a database? Feels like the cache process should just read the files and talk to the DB and skip grpc. Or we should use an actual message queue instead of files + grpc.

Reply:
I think in the case of downstream IPLD-ization in Postgres we would just consume the files directly as we know we want to consume everything- as quickly as possible- and will run that process in the same environment with direct access to the files. But if we want to be more selective about what we consume and consume it remotely than that's where this auxiliary streaming service comes into play, as I see it. @robert-zaremba may be able to back me up here.

Reply:
To actually answer your question: this would be useful primarily for clients but also potentially database caching in those cases where the caching wants to leverage this more selective interface and/or perform the caching remotely.

Reply:
Yes, it's for clients - eg data warehousing / ETL. We were discussing queues but didn't make a decision. Such a service will only need to expose a network port, without exposing filesystem.